Solve Deep Equal Problem From Day 9

unikuam opened the pull request:
1. Test case result is shown below:
![image](https://user-images.githubusercontent.com/59227893/80301449-c8788180-87c1-11ea-9dc6-b98ed6e8cf78.png)

2. The solution contains the following functions:

deepEqual() : default function, called initially
checkObjectEquality(): It checks if objects are equal or not recursively
hasSameType(): check if passed arguments have same type or not
hasNull(): if both arguments contain null then return true
hasSameDescriptors(): check if the property attributes match or not of provided object values
isObject(): check if object or not, additional check if it doesn't contain null 
checkArrayEquality(): check if two arrays equal or not, if any key contains object then check for object equality recursively
hasSameKeyLength(): check if provided parameters have the same number of keys or not

unikuam committed:
Solve Deep Equal Problem From Day 9

Added 30+ test cases.
Implemented the deep equality by convering all possible aspects.

ArfatSalman commented:
@unikuam Use back ticks for writing small snippets of code like `this`.

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+}
+
+function hasSameType(firstObj, secondObj) {
+  return typeof firstObj === typeof secondObj;
```
This would give true when one object is `null` and the other is an `object`

unikuam replied:
I missed this. I'll change the ```hasSameType()``` function accordingly.

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+    throw new TypeError(`Expected third argument as a boolean, got ${typeof matchDescriptors}`);
+  }
+  if (arguments.length <= 1) {
+    throw new TypeError(`Expected at least two arguments for compare, got ${arguments.length}`);
```
```suggestion
    throw new TypeError(`Expected at least two arguments for comparison, got ${arguments.length}`);
```

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+function checkObjectEquality(firstObj, secondObj, matchDescriptors) {
+  if (!hasSameType(firstObj, secondObj)) return false;
+  if (typeof firstObj !== 'object') {
+    if (typeof firstObj === 'function') return firstObj.toString() === secondObj.toString();
```
Put braces on single line if conditions.

unikuam replied:
Okay, I'll do this.

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+function checkObjectEquality(firstObj, secondObj, matchDescriptors) {
+  if (!hasSameType(firstObj, secondObj)) return false;
+  if (typeof firstObj !== 'object') {
+    if (typeof firstObj === 'function') return firstObj.toString() === secondObj.toString();
```
`(typeof firstObj === 'function') return firstObj.toString() === secondObj.toString();`

This would be true for most functions. You need to check them like objects.

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+}
+
+function isObject(obj) {
+  return typeof obj === 'object' && typeof obj !== null
```
This is wrong. 
```suggestion
  return typeof obj === 'object' && obj !== null
```

unikuam replied:
I wasn't aware of this error. It happened by mistake. I'll correct it. I didn't see ```typeof``` returning ```null``` in any case.

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+function hasSameDescriptors(firstObjDesc, secondObjDesc) {
+  for (const [key, firstObjDescValue] of Object.entries(firstObjDesc)) {
+    const secondObjDescValue = secondObjDesc[key];
+    if (Array.isArray(firstObjDescValue)) return checkArrayEquality(firstObjDescValue, secondObjDescValue);
```
I can't see that you have checked for descriptiors anywhere.

unikuam replied:
I used the following code:
```
if (matchDescriptors) {
      const firstObjDesc = Object.getOwnPropertyDescriptor(firstObj, key);
      const secondObjDesc = Object.getOwnPropertyDescriptor(secondObj, key);
      isEqual = isEqual && hasSameDescriptors(firstObjDesc, secondObjDesc);
    }
```
in ```checkObjectEquality()``` function. Please let me know if I am having any confusion here.

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+  for (const [key, firstObjDescValue] of Object.entries(firstObjDesc)) {
+    const secondObjDescValue = secondObjDesc[key];
+    if (Array.isArray(firstObjDescValue)) return checkArrayEquality(firstObjDescValue, secondObjDescValue);
+    else if (isObject(firstObjDescValue)) return checkObjectEquality(firstObjDescValue, secondObjDescValue, true);
```
Also, arrays and objects are the same thing in JS. You can have a single function checking both.

unikuam replied:
Okay, I'll check this. I think there was a case due to which I had to do it. I Will update here after checking the same again.

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+function hasSameKeyLength(firstObj, secondObj) {
+  const firstObjKeys = Object.getOwnPropertyNames(firstObj);
+  const secondObjKeys = Object.getOwnPropertyNames(secondObj);
+  if (firstObjKeys.length !== secondObjKeys.length) return false;
```
`firstObjKeys.length !== secondObjKeys.length` returns a boolean so you don't need to wrap it in an if-else.

unikuam replied:
You have mentioned this in live reviews many times. It won't happen again I hope

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+}
+
+function checkArrayEquality(firstObjDescValue, secondObjDescValue) {
+  for (let i = 0; i < firstObjDescValue.length; ++i) {
```
Use for-of loops when you know the objects are arrays.

unikuam replied:
Yaa, I should have used a for-of loop. Maybe, I was trying to pass the use of ```indexOf()```.

ArfatSalman commented on js-exercises/deep-equal/deepEqual.js:
```diff
@@ -0,0 +1,77 @@
+}
+
+function hasNull(...values) {
+  return values.every(el => el === null);
```
`hasNull` signifies that out of many things, some are nulls. In this case, the function will return true when every value is `null`. A better name should be `areNulls`.

unikuam replied:
Yes, I'll change it.

unikuam committed:
Fix Issues & Implement Changes Suggested by Reviewer

Function comparison like object is remaining here as I was using toString() earlier.
